### PR TITLE
Tn93: Fix filter output, update test data.

### PR DESCRIPTION
**Pull request by davebx**
FOR CONTRIBUTOR:
* [x] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [x] - License permits unrestricted use (educational + commercial)
* [x] - This PR updates an existing tool or tool collection

Fix combining clusters with TN-93.

### CODE REVIEW

File: tools/tn93/tn93_cluster.py [hunk 1]
```diff
@@ -41,27 +41,29 @@ def run_command(command):
...
+    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'background_msa.fa'))
+    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'background_msa.fa.bak'))
     with open(arguments.reference) as fh:
         for line in fh:
             if line[0] == '>':
                 _ref_seq_name = line[1:].split(' ')[0].strip()
                 break
     while True and threshold <= 1:
-        command = 'tn93-cluster -o clusters.json -t %g -a %s -c %s -m json -l %d -g %f reference_msa.fa' % (threshold, arguments.ambigs, arguments.cluster_type, arguments.overlap, arguments.fraction)
+        # input_stamp = run_command (task_runners['tn93'], ['-o', tn93_pairwise_calcs, '-s', ref_msa, '-t', "%g" % (threshold*2.0), query_compressed], tn93_pairwise_calcs, "filtering reference sequuences that are closer than %g to any query cluster" % (threshold*2.0))
```
Comment (bgruening) [inline tools/tn93/tn93_cluster.py:52]: remove?

File: tools/tn93/tn93_cluster.xml [hunk 2]
```diff
@@ -1,12 +1,12 @@
-<tool id="tn93_cluster" name="TN93 Cluster" version="@VERSION@">
+<tool id="tn93_cluster" name="TN93 Cluster" version="@WRAPPER_VERSION@">
     <description>sequences that lie within a specific distance of each other</description>
     <macros>
         <import>macros.xml</import>
+        <token name="@WRAPPER_VERSION@">@TOOL_VERSION@</token>
```
Comment (bgruening) [inline tools/tn93/tn93_cluster.xml:5]: why this extra token, you can have @TOOL_VERSION@ and optionally +galaxy0 also above, isn't it? WRAPPER_VERSION is not used twice.

File: tools/tn93/test-data/filter-out1.fasta [hunk 3]
```diff
@@ -13,3 +13,8 @@ ATGTTAGTTTTTCTTGTTTTATTGCCACTAGTCTCTAGTCAGTGTGTTAATCTTACAACCAGAACTCAATTACCCCCTGC
...
 
+
+epi_isl_1041403/hCoV-19/USA/NY-PRL-2021_02_08_05H08/2021
```
Comment (bernt-matthias) [inline tools/tn93/test-data/filter-out1.fasta:17]: This does not look like fasta. At least the `>` is missing.

File: tools/tn93/tn93_cluster.py [hunk 4]
```diff
@@ -41,27 +41,29 @@ def run_command(command):
...
-    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'reference_msa.fa'))
-    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'reference_msa.fa.bak'))
+    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'background_msa.fa'))
+    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'background_msa.fa.bak'))
     with open(arguments.reference) as fh:
         for line in fh:
             if line[0] == '>':
                 _ref_seq_name = line[1:].split(' ')[0].strip()
                 break
     while True and threshold <= 1:
```
Comment (bernt-matthias) [inline tools/tn93/tn93_cluster.py:51]: `True` seems redundant here.

File: tools/tn93/tn93_cluster.py [hunk 5]
```diff
@@ -41,27 +41,29 @@ def run_command(command):
 def main(arguments):
     threshold = arguments.threshold
     step = threshold * 0.25
-    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'reference_msa.fa'))
-    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'reference_msa.fa.bak'))
+    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'background_msa.fa'))
```
Comment (bernt-matthias) [inline tools/tn93/tn93_cluster.py:44]: Wouldn't linking be better?

File: tools/tn93/tn93_filter.xml [hunk 6]
```diff
@@ -1,20 +1,20 @@
...
+    </expand>
     <version_command><![CDATA[tn93 --version]]></version_command>
     <command detect_errors="exit_code"><![CDATA[
     cat '$reads' > filtered_msa.fa &&
     tn93 -o pairwise.csv -s filtered_msa.fa -t $threshold '$clusters' &&
     python '$__tool_directory__/tn93_filter.py' --reference '$reference'
         --distances pairwise.csv
         --reads filtered_msa.fa
-        --clusters '$clusters'
+        --background '$clusters'
```
Comment (bernt-matthias) [inline tools/tn93/tn93_filter.xml:17]: Is renaming really necessary, should we update the parameter label?

**Review by bernt-matthias — COMMENTED** [review]
Hi @davebx 

The PR misses context for me. I have no idea what problem is that you are trying to solve here. 

When looking at the tools they could certainly need more documentation as well.

Filter: 

- what is the `clusters` input doing?
- what is the meaning of "compressed"?
- in the changed test data sets it seems that the reference ends up in the output? Seems not expected for a filter, or?

cluster:

- the logic implemented in the python wrapper needs to be described.

**Comment by astrovsky01** [comment]
@bernt-matthias the original script this is running has updates which makes the current version outdated. We will fix the help text, but this is just updating Galaxy's python script to reflect the original tool's python script.

**Comment by wm75** [comment]
Seems that new help text is really unrelated to what the Galaxy tool is doing?

File: tools/tn93/tn93_filter.xml [hunk 7]
```diff
@@ -20,7 +19,7 @@
         <param name="reads" type="data" format="fasta" label="Aligned reads" />
         <param name="reference" type="data" format="fasta" label="Reference sequence" />
         <param name="clusters" type="data" format="fasta" label="FASTA file with compressed clusters" />
-        <param name="threshold" type="float" value="0.015" label="Pairwise distance threshold" />
+        <param name="threshold" type="float" value="0.015" min="0" max="1" label="Pairwise distance threshold" />
```
Comment (wm75) [inline tools/tn93/tn93_filter.xml:22]: The same change would be good in the other TN93 tools.

File: tools/tn93/tn93_cluster.py [hunk 8]
```diff
@@ -41,27 +40,22 @@ def run_command(command):
...
-    shutil.copy(arguments.input, os.path.join(os.getcwd(), 'reference_msa.fa.bak'))
     with open(arguments.reference) as fh:
         for line in fh:
             if line[0] == '>':
                 _ref_seq_name = line[1:].split(' ')[0].strip()
                 break
-    while True and threshold <= 1:
-        command = 'tn93-cluster -o clusters.json -t %g -a %s -c %s -m json -l %d -g %f reference_msa.fa' % (threshold, arguments.ambigs, arguments.cluster_type, arguments.overlap, arguments.fraction)
+    while threshold <= 1:
+        command = 'tn93-cluster -o clusters.json -t %g -a %s -c %s -m json -l %d -g %f %s' % (threshold, arguments.ambigs, arguments.cluster_type, arguments.overlap, arguments.fraction, arguments.input)
```
[the remark this answers is not in the document]
Comment (wm75) [reply tools/tn93/tn93_cluster.py:49]: So after having stared at this for a while my conclusion is that you're calling the underlying command line tool repeatedly with increasing thresholds until the number of clusters drops below the user specifed limit? Is that what's happening?

If so, then the wrapper should:
a) explain that
b) maybe offer two modes
 - a basic mode that just clusters based on a user-specified threshold
 - an iterative mode that dynamically adjusts the threshold until the specified number of clusters is reached
   in this mode there would be no reason to provide a threshold then, would there be?

**Comment by wm75** [comment]
> @bernt-matthias the original script this is running has updates which makes the current version outdated. We will fix the help text, but this is just updating Galaxy's python script to reflect the original tool's python script.

@astrovsky01 what's the "original tool's Python script"? Is there no new TN93 version coming with it?

File: tools/tn93/readreduce.xml [hunk 9]
```diff
@@ -1,11 +1,9 @@
-<tool id="tn93_readreduce" name="Merge matching reads" version="@VERSION@">
+<tool id="tn93_readreduce" name="Merge matching reads" version="@TOOL_VERSION@+galaxy0">
```
Comment (wm75) [inline tools/tn93/readreduce.xml:1]: `+galaxy0` is meant to indicate the first wrapper version for the `@TOOL_VERSION@` and is equivalent to no `+galaxyX` part at all.
So if you're not intending to bump `@TOOL_VERSION`, you should use `+galaxy1` here and in all other wrappers.

File: tools/tn93/tn93_cluster.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bgruening) [reply tools/tn93/tn93_cluster.py:49]: @davebx could you maybe improve the help section of this tool and mention the behavior.

**Review by bgruening — APPROVED** [review]

**Pull request merged**